Gems/Atom/RHI/Code/Tests/PageTileAllocatorTests.cpp

```diff
@@ -0,0 +1,170 @@
+        ASSERT_TRUE(allocator.GetUsedTileCount() == 0);
+        ASSERT_TRUE(allocator.IsPageFree());
+    }
+
+    TEST_F(PageTileAllocatorTest, SingleOutOfRangeAllocation_Failed)
+    {
+        RHI::PageTileAllocator allocator;
+        
+        uint32_t pageTileCount = 20;
+        allocator.Init(pageTileCount);
+
+        uint32_t allocated = 0;
+        uint32_t requested = 24;
+
+        AZStd::vector<RHI::PageTileSpan> tilesList = allocator.TryAllocate(requested, allocated);
+
+        ASSERT_TRUE(allocated == pageTileCount);
+        ASSERT_TRUE(tilesList.front().m_tileCount == allocated);
+        ASSERT_FALSE(allocator.IsPageFree());
+        
+        ASSERT_TRUE(allocator.GetFreeTileCount() == 0);
+        ASSERT_TRUE(allocator.GetUsedTileCount() == pageTileCount);
+
+        allocator.DeAllocate(tilesList);
+        
+        ASSERT_TRUE(allocator.GetFreeTileCount() == pageTileCount);
+        ASSERT_TRUE(allocator.GetUsedTileCount() == 0);
+        ASSERT_TRUE(allocator.IsPageFree());
+    }
+
+    TEST_F(PageTileAllocatorTest, RandomAllocationDeallocation_Success)
+    {
+        RHI::PageTileAllocator allocator;
+        
+        uint32_t pageTileCount = 30;
+        allocator.Init(pageTileCount);
+
+        uint32_t allocationCount = 100;
+
+        AZ::SimpleLcgRandom random(AZStd::GetTimeNowMicroSecond());
+
+        AZStd::vector<RHI::PageTileSpan> allocatedTilesList;
+
+        while (allocationCount-- != 0)
+        {
+            // 51% chance of doing an add. Biased towards adds so we fill up the allocator.
+            if ((random.GetRandom() % 100) <= 51 || allocatedTilesList.size() == 0)
+            {
+                uint32_t requested = random.GetRandom() % (pageTileCount + 10);
+                uint32_t allocated = 0;
+                AZStd::vector<RHI::PageTileSpan> tilesList = allocator.TryAllocate(requested, allocated);
+                allocatedTilesList.insert(AZStd::end(allocatedTilesList), AZStd::begin(tilesList), AZStd::end(tilesList));
+
+                ASSERT_TRUE(allocated <= requested);
+                ASSERT_TRUE(ValidateTilesNotOverlap(allocatedTilesList));
+                ASSERT_TRUE(GetTileCount(allocatedTilesList) == allocator.GetUsedTileCount());
+            }
+            else
+            {
+                // select some tile groups from the list
+                uint32_t count = random.GetRandom() % allocatedTilesList.size() + 1;
+                AZStd::vector<RHI::PageTileSpan> tilesToBeRemoved;
+                for (uint32_t i = 0; i < count; i++)
+                {
+                    uint32_t position = (random.GetRandom() % allocatedTilesList.size());
+                    tilesToBeRemoved.push_back(allocatedTilesList[position]);
+                    allocatedTilesList.erase(allocatedTilesList.begin() +position);
+                }
+                allocator.DeAllocate(tilesToBeRemoved);
+
+                ASSERT_TRUE(ValidateTilesNotOverlap(allocator.GetFreeList()));
```

Comment:
The merging is transparent to user. The benefit of the merging is to reduce allocation time and reduce the size of the free list. But it can be changed. The API doesn't promise merging.